File: UVA judge/524.cpp
#include <bits/stdc++.h>
#define all(x) x.begin(),x.end()
#define ff first
#define ss second
#define mt make_tuple
#define eb emplace_back

using namespace std;

typedef long long ll;
typedef vector<ll> vi;
typedef pair<ll,ll> pii;
typedef vector<pii> vii;
const ll INF = numeric_limits<ll>::max();

vi primes = {2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31};
vi container(17);
vector<vi> graph(17, vi());
ll n;


void search(ll index) {
  if (index == 1 and container[index] != 0) {

    auto repeat = 0;
    for (size_t i = 1; i <= n; i++) {
      if (container[i] != 0) {
        repeat++;
      }
    }

    if (repeat == n) {
      auto aux = container[1];
      
      std::cout << 1;
      while (aux != 1) {
        std::cout << ' ' << aux;
        aux = container[aux];
      }
      std::cout << '\n';
    }

    return;
  }

  for (auto i: graph[index]) {
    if (i <= n and container[index] == 0) {
      container[index] = i;
      search(i);
      container[index] = 0;
    }
    else
      break;
  }
}

int main()
{
  ios::sync_with_stdio(false);
  cin.tie(NULL);

  for (size_t i = 1; i < graph.size(); i++) {
    for (auto prime: primes)
      if (prime > i)
        graph[i].push_back(prime - i);
  }

  auto cas = 1;
  cin >> n;
  while (true) {

    for (size_t i = 1; i <= n; i++) {
      container[i] = 0;
    }

    std::cout << "Case " << cas << ':' << '\n';
    cas++;
    search(1);
    if (cin >> n)
      std::cout << '\n';
    else
      break;
  }


  return 0;
}

File: UVA judge/524.py
from sys import stdin, stdout, exit
from collections import deque
from math import inf

def main():
    cin = Scanner()
    primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31]
    container = [0 for i in range(17)]
    graph = [[] for i in range(17)]
    n = None

    for i in range(1, 17):
        for prime in primes:
            if prime > i:
                graph[i].append(prime - i)


    def search(index):
        if index is 1 and container[index] is not 0:

            repeat = 0
            for i in range(1, n + 1):
                if container[i] is not 0:
                    repeat += 1

            if repeat is n:
                print(1, end='')
                aux = container[1]
                while aux is not 1:
                    print(f' {aux}', end='')
                    aux = container[aux]
                print()

            return


        for i in graph[index]:
            if i <= n and container[index] is 0:
                container[index] = i
                search(i)
                container[index] = 0
            else:
                break

    case = 1
    while True:
        try:
            n = cin.next_int()
            if case is not 1 and n:
                print()
        except:
            exit(0)

        for i in range(1, n + 1):
            container[i] = 0

        print(f'Case {case}:')
        case += 1
        search(1)


class Scanner:

    def next_int(self):
        return int(self.next_line())

    def next_float(self):
        return float(self.next_line())

    def next_line(self):
        return stdin.readline().strip()

    def next_array(self, fun = str):
        line = self.next_line().split(' ')
        return line if fun is str else list(map(fun, line))


if __name__ == "__main__":
    main()
